### ISSUE TITLE
Gradient reversal layer

### ISSUE THREAD

**Issue by gpleiss** [pytorch/pytorch#1110]
It'd be nice to have a gradient reversal layer, like in Lua Torch. https://github.com/pytorch/pytorch/blob/master/torch/legacy/nn/GradientReversal.py

Or is there a more pytorch-esque way to accomplish gradient reversal?
Open to a PR?

**Comment by soumith** [comment]
Generally, you can achieve this with plain autograd, for example:

```
y = foo(x)
z = -1*y
```
z will flip the outputs and gradients.
Now if you ONLY want to reverse gradients, two ways:

1. Hooks. For example:
https://discuss.pytorch.org/t/updating-the-parameters-of-a-few-nodes-in-a-pre-trained-network-during-training/1221/5

2. Have a gradient reversal function:
Now, I am not sure if the use-cases for this are that great that we need a separate Function / Module for this. My understanding is that this is rare and when needed can be done with hooks or with autograd.

**Comment by apaszke** [comment]
I'd recommend 1, because as far as I remember autograd can sometimes rearrange nodes, and for these modifications to be correct, you need an assumption that they all compute true gradients. Use a hook, there are no assumptions about what's returned form them (just remember to never modify the grads received in the arguments in place)!

**Comment by gpleiss** [comment]
Thanks!
